### DISCUSSION TITLE
how did a normal function to be converted to bevy 'system'.

### DISCUSSION THREAD

**Discussion by cqlangyi** [bevyengine/bevy#5181] [Q&A] [answered]
with bevy 0.7

```rust
App::new()
.add_system(func1.before(func2))
...
```
with above code snippet, 'func1' be converted into a bevy 'system' automatically.

functions like 'before()', 'after()'... etc. they seem to be appended on to 'func1' some how, but don't know where from?

my question is: how did that happened?

any hint would be appreciated.

**Comment by tim-blackbird** [comment, marked as answer]
These methods come from an [extension trait](https://eli.thegreenplace.net/2022/rust-extension-traits-greppability-and-ides/) that is exported from the `bevy::prelude`.

**Comment by cqlangyi** [comment]
big thank for your answer. just what i'm expecting.
